ProjFS.Mac/PrjFSKextTests/VirtualizationRootsTests.mm

```diff
@@ -264,4 +271,119 @@ - (void)testRegisterProviderForPath_ExistingRecycledRoot
+    
+    
+    VirtualizationRootHandle repoRootHandle = InsertVirtualizationRoot_Locked(nullptr /* no client */, 0, repoRootVnode.get(), repoRootVnode->GetVid(), FsidInode{ repoRootVnode->GetMountPoint()->GetFsid(), repoRootVnode->GetInode() }, repoPath);
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(repoRootHandle));
+    
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, testFileVnode.get(), self->dummyVFSContext);
+    
+    XCTAssertEqual(foundRoot, RootHandle_None);
+}
+
+- (void)testFindForVnode_FileInUndetectedRoot
+{
+    const char* repoPath = "/Users/test/code/Repo";
+    const char* filePath = "/Users/test/code/Repo/file";
+    
+    shared_ptr<vnode> repoRootVnode = self->testMountPoint->CreateVnodeTree(repoPath, VDIR);
+    shared_ptr<vnode> testFileVnode = self->testMountPoint->CreateVnodeTree(filePath);
+
+    PrjFSVirtualizationRootXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    repoRootVnode->xattrs.insert(make_pair(PrjFSVirtualizationRootXAttrName, rootXattrData));
+
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, testFileVnode.get(), self->dummyVFSContext);
+    
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(foundRoot));
+}
+
+- (void)testFindForVnode_DetectRootRecycleThenFindRootForOtherFile
+{
+    const char* repoPath = "/Users/test/code/Repo";
+    const char* filePath = "/Users/test/code/Repo/file";
+    const char* otherFilePath = "/Users/test/code/Repo/some/otherfile";
+
+    shared_ptr<vnode> repoRootVnode = self->testMountPoint->CreateVnodeTree(repoPath, VDIR);
+    shared_ptr<vnode> testFileVnode = self->testMountPoint->CreateVnodeTree(filePath);
+
+    PrjFSVirtualizationRootXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    repoRootVnode->xattrs.insert(make_pair(PrjFSVirtualizationRootXAttrName, rootXattrData));
+
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, testFileVnode.get(), self->dummyVFSContext);
+
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(foundRoot));
+    
+    if (VirtualizationRoot_IsValidRootHandle(foundRoot))
+    {
+        XCTAssertEqual(s_virtualizationRoots[foundRoot].rootInode, repoRootVnode->GetInode());
+        
+        repoRootVnode->StartRecycling();
+        
+        shared_ptr<vnode> newRepoRootVnode = self->testMountPoint->CreateVnode(repoPath, VnodeCreationProperties { .inode = repoRootVnode->GetInode(), .type = VDIR });
```

Comment:
Why is `newRepoRootVnode` using the same inode as `repoRootVnode` with a different path?  Is this to test a hardlink scenario? (In general I'm not sure what how/why recycling `repoRootVnode` has any impact on `newRepoRootVnode`).

On a somewhat related note, I noticed that `Vnode_GetFsidAndInode` is ignoring the `useLinkIDForInode` parameter we're passing in.  Do you remember if there were plans to change what value is used for inode based on `useLinkIDForInode` ?

Reply:
> Why is `newRepoRootVnode` using the same inode as `repoRootVnode` with a different path?

As far as I can tell, and as far as intended, the paths are the same (`repoPath` in both instances):

```
shared_ptr<vnode> repoRootVnode = self->testMountPoint->CreateVnodeTree(repoPath, VDIR);
//…
shared_ptr<vnode> newRepoRootVnode = self->testMountPoint->CreateVnode(repoPath, VnodeCreationProperties { .inode = repoRootVnode->GetInode(), .type = VDIR });
```

> On a somewhat related note, I noticed that Vnode_GetFsidAndInode is ignoring the useLinkIDForInode parameter we're passing in. Do you remember if there were plans to change what value is used for inode based on useLinkIDForInode?

I'm planning to switch to the "real" `Vnode_GetFsidAndInode()` eventually and allow that to call down to `vnode_getattr()`. The distinction has so far not been important to the tests, and I've been doing the minimum reasonably possible to support the existing tests rather than building too far ahead and making these test PRs even bigger than they already are.

Reply:
>As far as I can tell, and as far as intended, the paths are the same

D'oh, I must have missed that in my earlier review, thanks!